libraries/Standard/src/primitive/resample.birch

```diff
@@ -273,14 +284,29 @@ function cumulative_weights(w:Real[_]) -> Real[_] {
  *
  * !!! note
  *     NaN log weights are treated as though `-inf`.
+ *
+ * This uses a numerically stable implementation that avoids over- and
+ * underflow, using a single pass over the data.
+ * It is based on:
+ *
+ * S. Nowozin (2016). Streaming Log-sum-exp Computation.
+ * http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html
  */
 function resample_reduce(w:Real[_]) -> (Real, Real) {
   if length(w) == 0 {
-    return (0.0, 0.0);
+    return (0.0, -inf);
   } else {
-    let mx <- -inf; // running maximum of log weights
-    let r <- 0.0; // running sum of non-maximum weights divided by maximum
-    let q <- 0.0; // running sum of square of non-maximum weights divided by maximum
+    // Running maximum of log weights
+    let mx <- -inf;
+    /*
+     * Running sum of non-maximum weights divided by maximum weight (r),
+     * and their squares (q).
+     * In contrast to Nowozin's implementation the maximum itself
+     * is not included in the sum.
+     * This avoids underflow for e.g. w = [1e-20, log(1e-20)].
+     */
+    let r <- 0.0;
+    let q <- 0.0;
     wn:Real;
     v:Real;
```

Comment:
Likewise, remove and use `let v <- ...` below.

Reply:
```suggestion
```